### PR TITLE
[JIT] ARM64/X64 - Add `IsVNNeverNegative`

### LINKED ISSUE #83248: [JIT] Recognize when length is never negative for `Span<T>::.ctor(void*, int32)`
I noticed this for the JIT intrinsic `BitOperations.PopCount`, for which the return value should be treated as never negative (since #64951).

```csharp
unsafe Span<byte> M0(byte* p, byte b) => new(p, b);
unsafe Span<byte> M2(byte* p, int b) => new(p, (byte)b);
unsafe Span<byte> M3(byte* p, uint mask) => new(p, BitOperations.PopCount(mask));
unsafe Span<byte> M4(byte* p, uint mask) => new(p, (byte)BitOperations.PopCount(mask));

```

```asm
// crossgen2 8.0.0-preview.3.23159.99+1b2eb12a711410a358bdfa93af017a34a929d4cf

C:M0(ulong,ubyte):System.Span`1[ubyte]:this:
; Emitting BLENDED_CODE for X64 CPU with AVX - Unix
       push     rax
						;; size=1 bbWeight=1 PerfScore 1.00
       mov      rax, rsi
       movzx    rdx, dl
						;; size=6 bbWeight=1 PerfScore 0.50
       add      rsp, 8
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25

C:M02(ulong,int):System.Span`1[ubyte]:this:
; Emitting BLENDED_CODE for X64 CPU with AVX - Unix
       push     rax
						;; size=1 bbWeight=1 PerfScore 1.00
       movzx    rdx, dl
       test     edx, edx
       jl       SHORT G_M62596_IG04
       mov      rax, rsi
						;; size=10 bbWeight=1 PerfScore 1.75
       add      rsp, 8
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
G_M62596_IG04:
       call     [System.ThrowHelper:ThrowArgumentOutOfRangeException()]
       int3     
						;; size=7 bbWeight=0 PerfScore 0.00

C:M1(ulong,uint):System.Span`1[ubyte]:this:
; Emitting BLENDED_CODE for X64 CPU with AVX - Unix
       push     rax
						;; size=1 bbWeight=1 PerfScore 1.00
       popcnt   edx, edx
       jl       SHORT G_M53218_IG04
       mov      rax, rsi
						;; size=9 bbWeight=1 PerfScore 3.25
       add      rsp, 8
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
G_M53218_IG04:
       call     [System.ThrowHelper:ThrowArgumentOutOfRangeException()]
       int3     
						;; size=7 bbWeight=0 PerfScore 0.00

C:M2(ulong,uint):System.Span`1[ubyte]:this:
; Emitting BLENDED_CODE for X64 CPU with AVX - Unix
       push     rax
						;; size=1 bbWeight=1 PerfScore 1.00
       popcnt   edx, edx
       jl       SHORT G_M40097_IG04
       mov      rax, rsi
						;; size=9 bbWeight=1 PerfScore 3.25
       add      rsp, 8
       ret      
						;; size=5 bbWeight=1 PerfScore 1.25
G_M40097_IG04:
       call     [System.ThrowHelper:ThrowArgumentOutOfRangeException()]
       int3     
						;; size=7 bbWeight=0 PerfScore 0.00
``

### PR DESCRIPTION
**Pull request by TIHan**
Resolves https://github.com/dotnet/runtime/issues/83248

### CODE REVIEW

File: src/coreclr/jit/valuenum.cpp [hunk 1]
```diff
@@ -5618,6 +5634,56 @@ bool ValueNumStore::IsVNInt32Constant(ValueNum vn)
...
+    }
+
+    VNFuncApp funcApp;
+    if (GetVNFunc(vn, &funcApp))
+    {
+        switch (funcApp.m_func)
+        {
+#ifdef FEATURE_HW_INTRINSICS
+#ifdef TARGET_XARCH
+            case VNF_HWI_POPCNT_PopCount:
```
Comment (EgorBo) [inline src/coreclr/jit/valuenum.cpp:5670]: Add
```
case NI_BMI1_TrailingZeroCount:
case NI_BMI1_X64_TrailingZeroCount:
case NI_LZCNT_LeadingZeroCount:
case NI_LZCNT_X64_LeadingZeroCount:
case NI_AdvSimd_LeadingZeroCount:
case NI_AdvSimd_LeadingSignCount:
case NI_ArmBase_LeadingZeroCount:
case NI_ArmBase_Arm64_LeadingZeroCount:
case NI_ArmBase_Arm64_LeadingSignCount:
```
?

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -5618,6 +5634,56 @@ bool ValueNumStore::IsVNInt32Constant(ValueNum vn)
     return TypeOfVN(vn) == TYP_INT;
 }
 
+bool ValueNumStore::IsVNNeverNegative(ValueNum vn)
+{
+    if (IsVNConstant(vn))
+    {
+        var_types vnTy = TypeOfVN(vn);
+        if (varTypeIsIntegral(vnTy))
```
Comment (AndyAyersMS) [inline src/coreclr/jit/valuenum.cpp:5642]: Probably faster if you check the type first thing and bail out quickly if it's not an integer.

Comment (AndyAyersMS) [reply src/coreclr/jit/valuenum.cpp:5642]: Ah you are checking it at the call site... so then why check again here?

Comment (TIHan) [reply src/coreclr/jit/valuenum.cpp:5666]: I could put an assert here for the type and we can always rely on the call-site to be responsible for what type it is.

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready. [Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=223446&view=ms.vss-build-web.run-extensions-tab)

File: src/coreclr/jit/valuenum.cpp [hunk 3]
```diff
@@ -4714,6 +4714,26 @@ ValueNum ValueNumStore::EvalUsingMathIdentity(var_types typ, VNFunc func, ValueN
                 {
                     resultVN = VNOneForType(typ);
                 }
+                else if (varTypeIsIntegralOrI(TypeOfVN(arg0VN)))
+                {
+                    ZeroVN = VNZeroForType(typ);
+                    if (genTreeOps(func) == GT_GE)
```
Comment (EgorBo) [inline src/coreclr/jit/valuenum.cpp:4720]: Can we handle VNF_GE_UN as well?

Comment (EgorBo) [reply src/coreclr/jit/valuenum.cpp:4720]: Also, do you handle `NN < 0` => false? (and its reversed version)

File: src/coreclr/jit/valuenum.cpp [hunk 4]
```diff
@@ -5618,6 +5658,64 @@ bool ValueNumStore::IsVNInt32Constant(ValueNum vn)
...
+#ifdef FEATURE_HW_INTRINSICS
+#ifdef TARGET_XARCH
+            case VNF_HWI_POPCNT_PopCount:
+            case VNF_HWI_POPCNT_X64_PopCount:
+            case VNF_HWI_LZCNT_LeadingZeroCount:
+            case VNF_HWI_LZCNT_X64_LeadingZeroCount:
+            case VNF_HWI_BMI1_TrailingZeroCount:
+            case VNF_HWI_BMI1_X64_TrailingZeroCount:
+                return true;
+#elif TARGET_ARM64
```
Comment (BruceForstall) [inline src/coreclr/jit/valuenum.cpp:5700]:
```suggestion
#elif defined(TARGET_ARM64)
```

File: src/coreclr/jit/valuenum.cpp (hunk 3 above)
Comment (TIHan) [reply src/coreclr/jit/valuenum.cpp:4720]: > Also, do you handle NN < 0 => false? (and its reversed version)
Yes, I handle it below.

Comment (TIHan) [reply src/coreclr/jit/valuenum.cpp:4720]: > Can we handle VNF_GE_UN as well?

What does `genTreeOps(VNF_GE_UN)` turn into? Would it just be `GT_GE` ?

Comment (EgorBo) [reply src/coreclr/jit/valuenum.cpp:4720]: > > Can we handle VNF_GE_UN as well?
> 
> What does `genTreeOps(VNF_GE_UN)` turn into? Would it just be `GT_GE` ?

It doesn't map into GenTreeOps, just check `func == VNF_GE_UN`

Comment (TIHan) [reply src/coreclr/jit/valuenum.cpp:4720]: Gotcha, then adding the checks should be good then.

Comment (TIHan) [reply src/coreclr/jit/valuenum.cpp:4720]: So it looks like we already handle `VNF_GE_UN` and others:
https://github.com/dotnet/runtime/blob/d6665c28dcbf7c71ec28969eb6382a8c6ade8f39/src/coreclr/jit/valuenum.cpp#L4852
https://github.com/dotnet/runtime/blob/d6665c28dcbf7c71ec28969eb6382a8c6ade8f39/src/coreclr/jit/valuenum.cpp#L4869

Therefore, we don't need to add additional checks for them.

**Review by EgorBo — APPROVED** [review]

**Pull request merged**